src/typeChecker/typeErrorChecker.ts

```diff
@@ -208,23 +212,38 @@ function typeCheckAndReturnType(node: tsEs.Node, context: Context, env: TypeEnvi
       return typeCheckAndReturnArrowFunctionType(node, context, env)
     }
     case 'FunctionDeclaration':
-      // Only identifiers are used as function params in Source
-      const params = node.params as tsEs.Identifier[]
+      if (node.id === null) {
+        // Block should not be reached since node.id is only null when function declaration
+        // is part of `export default function`, which is not used in Source
+        throw new TypecheckError(node, 'Function declaration should always have an identifier')
+      }
+
+      // Only identifiers/rest elements are used as function params in Source
+      const params = node.params as (tsEs.Identifier | tsEs.RestElement)[]
```

Comment:
Instead of type casting, is it possible to filter using type guards? You can then compare the length of the filtered array against that of the original array to throw an error if there is a mismatch (thus indicating that the invariant `Only identifiers/rest elements are used as function params in Source` is not true). This would help greatly in the event that this invariant is broken (whether on purpose or not) in the future.